radix-engine/src/engine/modules/execution_trace.rs

```diff
@@ -62,157 +110,194 @@ impl ExecutionTrace {
+            let bucket_id = call_data.bucket.0;
+
+            let frame = call_frames.last().expect("Current call frame not found");
+
+            if let Some(tree) = frame.owned_heap_nodes.get(&RENodeId::Bucket(bucket_id)) {
+                if let HeapRENode::Bucket(bucket_node) = &tree.root {
+                    track.vault_ops.push((
+                        actor.clone(),
+                        vault_id.clone(),
+                        VaultOp::Put(bucket_node.total_amount()),
+                    ));
+                }
             }
-        } else {
-            Err(RuntimeError::KernelError(KernelError::BucketNotFound(
-                bucket_id,
-            )))
         }
     }
 
-    fn handle_vault_take(
-        &mut self,
-        resource_address: &ResourceAddress,
-        component_id: &ComponentId,
+    fn handle_vault_take<'s, R: FeeReserve>(
+        track: &mut Track<'s, R>,
+        actor: &REActor,
         vault_id: &VaultId,
-        input: VaultTakeInput,
-    ) -> Result<(), RuntimeError> {
-        self.record_resource_change(resource_address, component_id, vault_id, -input.amount);
-        Ok(())
+        input: &ScryptoValue,
+    ) {
+        if let Ok(call_data) = scrypto_decode::<VaultTakeInput>(&input.raw) {
+            track.vault_ops.push((
+                actor.clone(),
+                vault_id.clone(),
+                VaultOp::Take(call_data.amount),
+            ));
+        }
     }
 
-    fn record_resource_change(
-        &mut self,
-        resource_address: &ResourceAddress,
-        component_id: &ComponentId,
+    fn handle_vault_lock_fee<'s, R: FeeReserve>(
+        track: &mut Track<'s, R>,
+        actor: &REActor,
         vault_id: &VaultId,
-        amount: Decimal,
     ) {
-        let component_changes = self
-            .resource_changes
-            .entry(component_id.clone())
-            .or_insert(HashMap::new());
-
-        let vault_change = component_changes
-            .entry(vault_id.clone())
-            .or_insert((resource_address.clone(), Decimal::zero()));
+        track
+            .vault_ops
+            .push((actor.clone(), vault_id.clone(), VaultOp::LockFee));
+    }
 
-        vault_change.1 += amount;
+    fn handle_vault_lock_contingent_fee<'s, R: FeeReserve>(
+        track: &mut Track<'s, R>,
+        actor: &REActor,
+        vault_id: &VaultId,
+    ) {
+        track
+            .vault_ops
+            .push((actor.clone(), vault_id.clone(), VaultOp::LockContingentFee));
     }
+}
+
+impl ExecutionTraceReceipt {
+    // TODO: is it better to derive resource changes from substate diff, instead of execution trace?
+    // The current approach relies on various runtime invariants.
+
+    pub fn new(
+        ops: Vec<(REActor, VaultId, VaultOp)>,
+        actual_fee_payments: HashMap<VaultId, Decimal>,
+        state_track: &mut AppStateTrack,
+        is_commit_success: bool,
```

Comment:
Does this correspond to `TransactionResult Commit/Reject` or `TransactionOutcome Success/Failure`? Should we use one of those names here for consistency?

Reply:
Yes, but I think it's consistent regards to naming:
- `Commit`
   * `Success`
   * `Failure`
- `Reject`

Reply:
Ahh, right `is_Commit->Success`, makes sense